vsomeip-sys/build.rs

```diff
@@ -0,0 +1,264 @@
+/********************************************************************************
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use decompress::ExtractOptsBuilder;
+use reqwest::blocking::Client;
+use std::error::Error;
+use std::fs::File;
+use std::io::{BufRead, BufReader, BufWriter, Write};
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::time::Duration;
+use std::{env, fs};
+
+const VSOMEIP_TAGGED_RELEASE_BASE: &str = "https://github.com/COVESA/vsomeip/archive/refs/tags/";
+const VSOMEIP_VERSION_ARCHIVE: &str = "3.4.10.tar.gz";
+
+fn main() -> miette::Result<()> {
+    let out_dir = env::var_os("OUT_DIR").unwrap();
+
+    let vsomeip_interface_path = {
+        // here we allow bringing the user's own vsomeip interface to bind against if they so wish
+        // note that this binding is configured to work with the tagged release above of VSOMEIP_VERSION_ARCHIVE
+        let user_supplied_vsomeip_include_path = env::var("VSOMEIP_INCLUDE_PATH");
+        if let Ok(user_supplied_vsomeip_include_path) = user_supplied_vsomeip_include_path {
+            PathBuf::from(user_supplied_vsomeip_include_path)
+        } else {
+            let vsomeip_archive_dest = Path::new(&out_dir).join("vsomeip").join("vsomeip.tar.gz");
+            let vsomeip_archive_url =
+                format!("{VSOMEIP_TAGGED_RELEASE_BASE}{VSOMEIP_VERSION_ARCHIVE}");
+            let vsomeip_source_folder = Path::new(&out_dir).join("vsomeip").join("vsomeip-src");
+            download_and_write_file(&vsomeip_archive_url, &vsomeip_archive_dest)
+                .expect("Unable to download released archive");
+            decompress::decompress(
+                vsomeip_archive_dest,
+                vsomeip_source_folder.clone(),
+                &ExtractOptsBuilder::default().strip(1).build().unwrap(),
+            )
+            .expect("Unable to extract tar.gz");
+            vsomeip_source_folder.join("interface")
```

Comment:
FMPOV we should rely on the user always providing the path

Reply:
Yeah we chatted about this. I'll probably have to make this change for the reasons you outlined above.

Reply:
Added the ability to compile the downloaded vsomeip library for the user in the latest commit. Looking for feedback :slightly_smiling_face: